File: hushh-webapp/ios/App/App/Plugins/HushhStreamPlugin.swift
import Foundation
import Capacitor

/**
 * HushhStream: a generic streaming HTTP transport for the WebView.
 *
 * WKWebView buffers a `fetch()` response body until the request completes,
 * and the app routes native calls through CapacitorHttp, which returns whole
 * bodies; a server-sent event stream therefore reaches the page only when it
 * has ended. This plugin performs the request with a delegate `URLSession`
 * and forwards the raw response bytes as they arrive (`hushhStreamEvent`,
 * `type: "chunk"`, base64), then one `end` event. It carries bytes only: no
 * SSE parsing, no JSON, no retry. Several streams may run at once, each
 * addressed by the caller's `streamId`.
 *
 * `open` resolves when the response headers arrive (`{status, headers}`), so
 * a non-2xx status is known before any body byte; the body still streams so
 * the JavaScript side can read the error payload. The backend URL for an
 * app-relative path resolves the way every other plugin resolves it
 * (`HushhProxyClient`). Mirrors the URLSession pattern of `KaiPlugin`.
 */
@objc(HushhStreamPlugin)
public class HushhStreamPlugin: CAPPlugin, CAPBridgedPlugin, URLSessionDataDelegate {

    private let TAG = "HushhStreamPlugin"

    // MARK: - CAPBridgedPlugin Protocol (MUST be declared before any other properties)
    public let identifier = "HushhStreamPlugin"
    public let jsName = "HushhStream"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancel", returnType: CAPPluginReturnPromise),
    ]

    private static let eventName = "hushhStreamEvent"
    private static let requestTimeout: TimeInterval = 180
    private static let resourceTimeout: TimeInterval = 900

    private final class StreamState {
        let streamId: String
        var openCall: CAPPluginCall?
        var task: URLSessionDataTask?
        var status: Int = 0
        var cancelled = false

        init(streamId: String, openCall: CAPPluginCall) {
            self.streamId = streamId
            self.openCall = openCall
        }
    }

    private lazy var session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = HushhStreamPlugin.requestTimeout
        config.timeoutIntervalForResource = HushhStreamPlugin.resourceTimeout
        config.requestCachePolicy = .reloadIgnoringLocalCacheData
        return URLSession(configuration: config, delegate: self, delegateQueue: nil)
    }()

    private let registryLock = NSLock()
    private var byTaskIdentifier: [Int: StreamState] = [:]
    private var byStreamId: [String: StreamState] = [:]

    // MARK: - Registry

    private func register(_ state: StreamState, task: URLSessionDataTask) {
        registryLock.lock()
        state.task = task
        byTaskIdentifier[task.taskIdentifier] = state
        byStreamId[state.streamId] = state
        registryLock.unlock()
    }

    private func state(for task: URLSessionTask) -> StreamState? {
        registryLock.lock()
        defer { registryLock.unlock() }
        return byTaskIdentifier[task.taskIdentifier]
    }

    private func state(for streamId: String) -> StreamState? {
        registryLock.lock()
        defer { registryLock.unlock() }
        return byStreamId[streamId]
    }

    private func remove(_ state: StreamState) {
        registryLock.lock()
        if let task = state.task {
            byTaskIdentifier.removeValue(forKey: task.taskIdentifier)
        }
        byStreamId.removeValue(forKey: state.streamId)
        registryLock.unlock()
    }

    private func emit(_ data: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.notifyListeners(HushhStreamPlugin.eventName, data: data)
        }
    }

    private func emitEnd(_ state: StreamState, error: String?, code: String?) {
        var payload: [String: Any] = ["streamId": state.streamId, "type": "end"]
        if let error = error { payload["error"] = error }
        if let code = code { payload["code"] = code }
        emit(payload)
    }

    // MARK: - Plugin Methods

    @objc func open(_ call: CAPPluginCall) {
        guard let streamId = call.getString("streamId")?.trimmingCharacters(in: .whitespacesAndNewlines),
              !streamId.isEmpty else {
            call.reject("Missing streamId")
            return
        }
        guard let path = call.getString("path"), !path.isEmpty else {
            call.reject("Missing path")
            return
        }
        if state(for: streamId) != nil {
            call.reject("A stream with this id is already open", "HUSHH_STREAM_BUSY")
            return
        }

        let urlString: String
        if path.lowercased().hasPrefix("http://") || path.lowercased().hasPrefix("https://") {
            urlString = path
        } else {
            let backendUrl = HushhProxyClient.resolveBackendUrl(call: call, plugin: self, jsName: jsName)
            guard !backendUrl.isEmpty else {
                call.reject("Backend URL is not configured", "HUSHH_STREAM_NO_BACKEND")
                return
            }
            urlString = "\(backendUrl)\(path.hasPrefix("/") ? path : "/\(path)")"
        }
        guard let url = URL(string: urlString) else {
            call.reject("Invalid URL: \(urlString)")
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = (call.getString("method") ?? "GET").uppercased()
        if let headers = call.getObject("headers") {
            for (key, value) in headers {
                if let text = value as? String {
                    request.setValue(text, forHTTPHeaderField: key)
                }
            }
        }
        if let body = call.getString("body"), request.httpMethod != "GET" {
            request.httpBody = body.data(using: .utf8)
        }

        // The call is held by the stream state and settled when the response
        // headers arrive (the same late-resolve pattern as KaiPlugin).
        let state = StreamState(streamId: streamId, openCall: call)
        let task = session.dataTask(with: request)
        register(state, task: task)
        task.resume()
    }

    @objc func cancel(_ call: CAPPluginCall) {
        guard let streamId = call.getString("streamId"), let state = state(for: streamId) else {
            call.resolve(["cancelled": false])
            return
        }
        state.cancelled = true
        state.task?.cancel()
        call.resolve(["cancelled": true])
    }

    // MARK: - URLSessionDataDelegate

    public func urlSession(
        _ session: URLSession,
        dataTask: URLSessionDataTask,
        didReceive response: URLResponse,
        completionHandler: @escaping (URLSession.ResponseDisposition) -> Void
    ) {
        guard let state = state(for: dataTask) else {
            completionHandler(.cancel)
            return
        }
        let http = response as? HTTPURLResponse
        let status = http?.statusCode ?? 0
        state.status = status
        var headers: [String: String] = [:]
        if let http = http {
            for (key, value) in http.allHeaderFields {
                if let name = key as? String, let text = value as? String {
                    headers[name.lowercased()] = text
                }
            }
        }
        let openCall = state.openCall
        state.openCall = nil
        DispatchQueue.main.async {
            openCall?.resolve(["status": status, "headers": headers])
        }
        completionHandler(.allow)
    }

    public func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        guard let state = state(for: dataTask), !state.cancelled, !data.isEmpty else { return }
        emit([
            "streamId": state.streamId,
            "type": "chunk",
            "base64": data.base64EncodedString(),
        ])
    }

    public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        guard let state = state(for: task) else { return }
        remove(state)

        // A failure before the headers arrived settles the pending `open`.
        if let openCall = state.openCall {
            state.openCall = nil
            DispatchQueue.main.async {
                let code = state.cancelled ? "CANCELLED" : HushhStreamPlugin.errorCode(error)
                openCall.reject(error?.localizedDescription ?? "Stream request failed", code)
            }
        }

        if state.cancelled {
            emitEnd(state, error: "Cancelled", code: "CANCELLED")
            return
        }
        if let error = error {
            emitEnd(state, error: error.localizedDescription, code: HushhStreamPlugin.errorCode(error))
            return
        }
        emitEnd(state, error: nil, code: nil)
    }

    private static func errorCode(_ error: Error?) -> String {
        guard let error = error as NSError? else { return "NETWORK" }
        switch error.code {
        case NSURLErrorCancelled: return "CANCELLED"
        case NSURLErrorTimedOut: return "TIMEOUT"
        default: return "NETWORK"
        }
    }
}

File: hushh-webapp/android/app/src/main/java/com/hussh/app/plugins/HushhStream/HushhStreamPlugin.kt
package com.hussh.app.plugins.HushhStream

import android.util.Base64
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.hussh.app.plugins.shared.BackendUrl
import okhttp3.Call
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/**
 * HushhStream: a generic streaming HTTP transport for the WebView.
 *
 * The app routes native calls through CapacitorHttp, which returns whole
 * bodies, so a server-sent event stream reaches the page only when it has
 * ended. This plugin performs the request with OkHttp on a worker thread and
 * forwards the raw response bytes as they arrive (`hushhStreamEvent`,
 * `type: "chunk"`, base64), then one `end` event. Bytes only: no SSE
 * parsing, no JSON, no retry. Several streams may run at once, each
 * addressed by the caller's `streamId`. Port of the iOS HushhStreamPlugin;
 * the OkHttp pattern is KaiPlugin's.
 */
@CapacitorPlugin(name = "HushhStream")
class HushhStreamPlugin : Plugin() {

    private val TAG = "HushhStreamPlugin"

    companion object {
        private const val EVENT_NAME = "hushhStreamEvent"
        private const val CHUNK_BYTES = 8 * 1024
    }

    private class StreamState(val streamId: String) {
        @Volatile var call: Call? = null
        @Volatile var cancelled = false
    }

    private val streams = ConcurrentHashMap<String, StreamState>()

    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .readTimeout(180, TimeUnit.SECONDS)
        // A stream can legitimately run longer than a fixed wall clock budget.
        .callTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    private fun emit(payload: JSObject) {
        activity.runOnUiThread { notifyListeners(EVENT_NAME, payload) }
    }

    private fun emitEnd(state: StreamState, error: String?, code: String?) {
        val payload = JSObject().put("streamId", state.streamId).put("type", "end")
        if (error != null) payload.put("error", error)
        if (code != null) payload.put("code", code)
        emit(payload)
    }

    @PluginMethod
    fun open(call: PluginCall) {
        val streamId = call.getString("streamId")?.trim().orEmpty()
        if (streamId.isEmpty()) {
            call.reject("Missing streamId")
            return
        }
        val path = call.getString("path").orEmpty()
        if (path.isEmpty()) {
            call.reject("Missing path")
            return
        }
        if (streams.containsKey(streamId)) {
            call.reject("A stream with this id is already open", "HUSHH_STREAM_BUSY")
            return
        }

        val url = if (path.startsWith("http://", true) || path.startsWith("https://", true)) {
            path
        } else {
            val backendUrl = BackendUrl.resolve(bridge, call, "HushhStream")
            if (backendUrl.isEmpty()) {
                call.reject("Backend URL is not configured", "HUSHH_STREAM_NO_BACKEND")
                return
            }
            backendUrl + (if (path.startsWith("/")) path else "/$path")
        }

        val method = (call.getString("method") ?: "GET").uppercase()
        val builder = Request.Builder().url(url)
        var contentType = "application/json"
        call.getObject("headers")?.let { headers ->
            val keys = headers.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                val value = headers.optString(key, "")
                if (value.isNotEmpty()) {
                    builder.addHeader(key, value)
                    if (key.equals("Content-Type", ignoreCase = true)) contentType = value
                }
            }
        }
        val body = call.getString("body")
        if (method == "GET") {
            builder.get()
        } else {
            builder.method(method, (body ?: "").toRequestBody(contentType.toMediaType()))
        }

        val state = StreamState(streamId)
        val httpCall = httpClient.newCall(builder.build())
        state.call = httpCall
        streams[streamId] = state

        Thread {
            var settledOpen = false
            try {
                httpCall.execute().use { response ->
                    val headers = JSObject()
                    for (name in response.headers.names()) {
                        headers.put(name.lowercase(), response.headers[name] ?: "")
                    }
                    val opened = JSObject().put("status", response.code).put("headers", headers)
                    settledOpen = true
                    activity.runOnUiThread { call.resolve(opened) }

                    val stream = response.body?.byteStream()
                    if (stream != null) {
                        val buffer = ByteArray(CHUNK_BYTES)
                        while (!state.cancelled) {
                            val read = stream.read(buffer)
                            if (read < 0) break
                            if (read == 0) continue
                            val encoded = Base64.encodeToString(buffer, 0, read, Base64.NO_WRAP)
                            emit(JSObject().put("streamId", streamId).put("type", "chunk").put("base64", encoded))
                        }
                    }
                }
                if (state.cancelled) emitEnd(state, "Cancelled", "CANCELLED") else emitEnd(state, null, null)
            } catch (e: IOException) {
                val code = if (state.cancelled) "CANCELLED" else if (e is java.net.SocketTimeoutException) "TIMEOUT" else "NETWORK"
                if (!settledOpen) {
                    activity.runOnUiThread { call.reject(e.message ?: "Stream request failed", code) }
                }
                emitEnd(state, e.message ?: "Stream failed", code)
            } catch (e: Exception) {
                android.util.Log.e(TAG, "open error", e)
                if (!settledOpen) {
                    activity.runOnUiThread { call.reject(e.message ?: "Stream request failed", "NETWORK") }
                }
                emitEnd(state, e.message ?: "Stream failed", "NETWORK")
            } finally {
                streams.remove(streamId)
            }
        }.start()
    }

    @PluginMethod
    fun cancel(call: PluginCall) {
        val streamId = call.getString("streamId").orEmpty()
        val state = streams[streamId]
        if (state == null) {
            call.resolve(JSObject().put("cancelled", false))
            return
        }
        state.cancelled = true
        state.call?.cancel()
        call.resolve(JSObject().put("cancelled", true))
    }
}
